### PR TITLE
feat(moss-td): opt-in torch.compile for the Whisper encoder

### PR DESCRIPTION
**Pull request by draftbk**
Opt-in `torch.compile(mode="reduce-overhead")` for the MOSS-Transcribe-Diarize Whisper encoder (#924), as an alternative encoder acceleration to the CUDA graph from #973 — the two are mutually exclusive, not stacked (reduce-overhead already fuses *and* CUDA-graphs the encoder). Enabling the new `encoder_torch_compile` stage arg (off by default) makes the encoder use compile **instead of** the CUDA graph — when the flag is on, the encoder CUDA graph is not built. Both mechanisms share one `encoder_chunk_buckets` arg (default `1..8`, ~4 min of audio); `dynamic=False` compiles per exact chunk count, so counts outside the warmed buckets fall back to eager.

When to use: encoder-bound, high-concurrency short-audio serving (see the numbers below). Leaving it off keeps the default #973 CUDA graph.

## Results

Test env: A100-40GB · bf16 · torch 2.11.0+cu130 · latest `main` (includes #972 CPU encoder cache and #973 encoder CUDA graph) · movies800 (96 samples) · greedy · fresh server per run · `mem_fraction_static=0.5`. 3 runs each; all 96/96 completed. LLM CUDA graph on for every row — only the encoder path differs.

### Isolated encoder forward (n=1, CUDA-event timed, ~19 warm samples)

| encoder path | ms | vs eager |
|---|---|---|
| eager | 10.32 (sd 0.12) | — |
| CUDA graph (#973) | 9.72 (sd 0.004) | −5.8% |
| compile (fusion only) | 8.59 (sd 0.13) | −16.7% |
| compile reduce-overhead | 8.20 (sd 0.13) | **−20.5%** |

### End-to-end throughput (qps), 3 runs each

| conc | eager | CUDA graph | compile | compile RO |
|---|---|---|---|---|
| 1 | 2.27 / 2.34 / 2.34 | 2.35 / 2.34 / 2.34 | 2.35 / 2.36 / 2.34 | 2.36 / 2.34 / 2.36 |
| 4 | 7.43 / 7.35 / 7.42 | 7.56 / 7.43 / 7.54 | 7.56 / 7.57 / 7.57 | 7.61 / 7.54 / 7.69 |
| 8 | 11.62 / 11.62 / 11.58 | 11.88 / 11.93 / 11.73 | 11.98 / 11.94 / 11.85 | 12.03 / 12.05 / 12.04 |
| 16 | 16.33 / 16.34 / 16.62 | 16.79 / 16.87 / 16.71 | 17.11 / 16.78 / 17.10 | 17.30 / 17.08 / 17.27 |

Means:

| conc | eager | CUDA graph | compile | compile RO | RO vs eager | RO vs CUDA graph |
|---|---|---|---|---|---|---|
| 1 | 2.315 | 2.345 | 2.350 | 2.353 | +1.6% | +0.3% |
| 4 | 7.401 | 7.510 | 7.565 | 7.614 | +2.9% | +1.4% |
| 8 | 11.606 | 11.847 | 11.921 | 12.040 | +3.7% | +1.6% |
| 16 | 16.426 | 16.788 | 16.999 | **17.218** | **+4.8%** | **+2.6%** |

CER: eager and CUDA graph are numerically exact; compile drifts ~0.05–0.2 pp from fusion numerics.

### CODE REVIEW

**Comment by Ccyest** [comment]
please rebase and run a benchmark comparison against upstream main. Since CUDA Graph just merged.

File: sglang_omni/models/moss_transcribe_diarize/sglang_model.py [hunk 1]
```diff
@@ -99,6 +100,22 @@ def init_encoder_cache(self, max_bytes: int) -> None:
...
+        p = next(self.whisper_encoder.parameters())
+        frames = int(cfg.max_source_positions) * 2
+        pos = torch.arange((frames - 1) // 2 + 1, device=p.device, dtype=torch.long)
+        with torch.no_grad():
+            for n in buckets:
+                feats = torch.zeros(
+                    n, int(cfg.num_mel_bins), frames, device=p.device, dtype=p.dtype
+                )
+                self.encoder_runner(feats, pos, None)
+
```
Comment (Ccyest) [inline sglang_omni/models/moss_transcribe_diarize/sglang_model.py]: The default buckets (1,2,3,4) only cover roughly ≤2 minutes of audio; 
No eager fallback on long audio.

Suggested fix: similar to cudagraph, pre warm-up small bucket size like [1,9), which covers 4 minutes. And fallback to eager if exceeds it.

File: sglang_omni/models/moss_transcribe_diarize/sglang_model.py [hunk 2]
```diff
@@ -75,6 +75,7 @@ def __init__(
         super().__init__()
         self.config = config
         self.whisper_encoder = WhisperEncoder(config.audio_config, quant_config)
+        self.encoder_runner = self.whisper_encoder
```
Comment (Ccyest) [inline sglang_omni/models/moss_transcribe_diarize/sglang_model.py]: **rebase related from claude code**
Derived from the root-cause issue above (both mechanisms must coexist after the rebase). self.encoder_runner = self.whisper_encoder goes through nn.Module.__setattr__ and registers the same encoder under a second name in _modules; after compilation, the OptimizedModule hangs in as a third copy (encoder_runner._orig_mod.*). Consequence: state_dict() / named_parameters() now expose two to three sets of encoder parameter names. Today's load_weights happens to be unaffected (it only maps checkpoint names onto whisper_encoder.*, and loading happens before compilation — I traced the params_dict lookups), but anything that later relies on state_dict (checkpoint saving, weight refit, parameter accounting) will see duplicate entries. This is a landmine. Main's #973 already established the cleaner pattern: self._encoder_graph_runner = None is a plain attribute (None is not a Module, so nothing is registered), with an explicit if/else at the call site.

**Fix: **:
```
# __init__
self._compiled_encoder = None  # plain attr; not a submodule

# compile_encoder
self._compiled_encoder = torch.compile(self.whisper_encoder, dynamic=False)

# call site (post-rebase, precedence made explicit)
if self._encoder_graph_runner is not None:
    features = self._encoder_graph_runner.run(...)
elif self._compiled_encoder is not None:
    features = self._compiled_encoder(...)
else:
    features = self.whisper_encoder(...)
```

File: sglang_omni/models/moss_transcribe_diarize/stages.py [hunk 3]
```diff
@@ -90,6 +90,8 @@ def create_sglang_moss_transcribe_diarize_executor(
     mem_fraction_static: float | None = 0.80,
     mm_embedding_cache_size_bytes: int = 0,
     encoder_cache_size_bytes: int = 0,
+    encoder_torch_compile: bool = False,
+    encoder_compile_buckets: Tuple[int, ...] = (1, 2, 3, 4),
```
Comment (Ccyest) [inline sglang_omni/models/moss_transcribe_diarize/stages.py]: **Rebase related from claude code:**
Derived from the root-cause issue. After the rebase, the same physical quantity — the per-forward chunk count of the encoder — ends up with two factory arguments: this PR's encoder_compile_buckets: Tuple[int, ...] = (1, 2, 3, 4) and #973's encoder_graph_chunk_buckets: list[int] | None = None (expanding to [1..8] by default). One is a tuple, the other a list; default coverage goes to 4 vs 8; and #973's capture() filters invalid values (int(x) >= 1), deduplicates, and skips buckets when VRAM is low, whereas compile_encoder takes the buckets verbatim — passing 0 warms an empty-batch graph, and duplicates re-run warmup. Two nearly-identical knobs in the same function signature is guaranteed user confusion.

Suggestion:
Unify into a single encoder_chunk_buckets argument shared by both mechanisms (their physical meaning of a bucket is identical), and consolidate the {int(b) for b in buckets if int(b) >= 1} normalization in one place.

Comment (draftbk) [reply sglang_omni/models/moss_transcribe_diarize/stages.py]: Done

File: sglang_omni/models/moss_transcribe_diarize/sglang_model.py (hunk 1 above)
Comment (draftbk) [reply sglang_omni/models/moss_transcribe_diarize/sglang_model.py]: Done

File: sglang_omni/models/moss_transcribe_diarize/sglang_model.py (hunk 2 above)
Comment (draftbk) [reply sglang_omni/models/moss_transcribe_diarize/sglang_model.py]: Fixed.

**Comment by draftbk** [comment]
> please rebase and run a benchmark comparison against upstream main. Since CUDA Graph just merged.

Done, rebased onto latest main and re-ran the eval, 4-way comparison is in the description.

**Review by Ccyest — APPROVED** [review]
